Proper np array object to string conversion

RyanK1m opened the pull request:
tostring() function return bytes which cannot be properly decoded with decode('utf-8') in markdown_and_sanitize function.

```python
def markdown_and_sanitize(markdown_string):
  if isinstance(markdown_string, six.binary_type):
  markdown_string = markdown_string.decode('utf-8')
```
decoded **markdown_string** contains 3 NULL padding per every character and 8 NULL trailing padding at the end of the string. And **markdown module** cannot propery handle this null padded string, hence the #222 and other erronous markdown behavior.

This pull request change **"tostring"** function to np.array_str which return a plain string without any null padding. 

**markdown module** can handle this plain string without any problem, so that fixes #222 and other markdown problems.

wchargin commented:
It's not clear to me what problem this solves. TensorFlow, numpy, and UTF-8 all handle strings with NUL characters just fine:
```
print(repr(tf.make_ndarray(tf.make_tensor_proto('test\0\0\0there', dtype=tf.string)).astype(np.dtype(str)).tostring().decode('utf-8')))
# u'test\x00\x00\x00there'
```

Please add a test case that _fails_ before your change and _passes_ after your change, so that we can see exactly what the problem is.

wchargin commented:
It's not clear to me what problem this solves. TensorFlow, numpy, and UTF-8 all handle strings with NUL characters just fine:
```py
print(repr(tf.make_ndarray(tf.make_tensor_proto('test\0\0\0there', dtype=tf.string)).astype(np.dtype(str)).tostring().decode('utf-8')))
# u'test\x00\x00\x00there'
```

Please add a test case that _fails_ before your change and _passes_ after your change, so that we can see exactly what the problem is.

wchargin commented:
One more point of confusion: you mention that the `markdown` module can't handle strings with NUL characters, but they seem just fine to me:
```py
>>> import markdown
>>> markdown.markdown('\0\0\0hello\0world\0\0\0', extensions=['markdown.extensions.tables'])
u'<p>\x00\x00\x00hello\x00world\x00\x00\x00</p>'
```

Likewise `bleach`:
```py
>>> import bleach
>>> bleach.clean('<p>\0\0\0hello\0world\0\0\0</p>')
u'&lt;p&gt;helloworld&lt;/p&gt;'
```

Providing a test case as I requested above would elucidate all these issues.

RyanK1m commented:
I'll add test case as soon as possible. :)

It seems markdown module doesn't work with zero padded string when you want to use table extension. And tensorboard code uses this extension.

Here's the example

```python
import markdown
markdown_string = """test1|test2\n---|---\nvalue1|value2"""
markdown_zero_padded_string = ""
for char in markdown_string:
  markdown_zero_padded_string += char +'\0'
string_html_with_zero_padding = markdown.markdown(
      markdown_zero_padded_string, extensions=['markdown.extensions.tables'])
string_html = markdown.markdown(
      markdown_string, extensions=['markdown.extensions.tables'])
print('-------normal string-------')
print(string_html)
print('-------zero padded string-------')
print(string_html_with_zero_padding)
```

result
```
-------normal string-------
<table>
<thead>
<tr>
<th>test1</th>
<th>test2</th>
</tr>
</thead>
<tbody>
<tr>
<td>value1</td>
<td>value2</td>
</tr>
</tbody>
</table>
-------zero padded string-------
<p>test1|test2
---|---
value1|value2</p>
```

If I want to use Text summary with table, it gives me above raw string.

By the way, current test code testing "markdown_and_sanitize" with table works fine because test string itself doesn't contain zero padding. 

```python
table = textwrap.dedent("""\
    An | Example | Table
    --- | --- | ---
    A | B | C
    1 | 2 | 3
    """)

    table_converted = textwrap.dedent("""\
    <table>
    <thead>
    <tr>
    <th>An</th>
    <th>Example</th>
    <th>Table</th>
    </tr>
    </thead>
    <tbody>
    <tr>
    <td>A</td>
    <td>B</td>
    <td>C</td>
    </tr>
    <tr>
    <td>1</td>
    <td>2</td>
    <td>3</td>
    </tr>
    </tbody>
    </table>""")

self.assertEqual(text_plugin.markdown_and_sanitize(table), table_converted)
```

wchargin commented:
Ah, thanks for clarifying. A minimal example is that `a|b\n-|-\nc|d` is parsed as a table but `a|b\n\0-|-\nc|d` is not.

This explains where the Markdown module isn't behaving as you expected. But the behavior of the Markdown module makes sense: `\0--- | ---` is not a row specification, so this shouldn't be parsed as a table. The root cause comes earlier: why does the string that you're trying to use as a table have NUL characters in it?

First: what is the rank of string tensor that you're passing to `tf.summary.text`? (Or, rather, what is the lowest rank for which you're able to reproduce surprising behavior?)

Now, consider the following statements:
  - the contents of the string tensor that you're passing to `tf.summary.text` contains NUL characters;
  - the event files on disk contain NUL characters within your text;
  - the protobufs read into memory by TensorBoard's event accumulator contain NUL characters;
  - the tensor protos returned by `self._multiplexer.Tensors` in the text plugin contain NUL characters;
  - the result of `tf.make_ndarray(event.tensor_proto)` contains NUL characters;
  - any argument to `markdown_and_sanitize` contains NUL characters.

As far as I can tell, the above statements *should* all be equivalent, and furthermore if they are all equivalent then TensorBoard's behavior is correct.

> I'll add test case as soon as possible.

Great; no rush. :-)

RyanK1m commented:
The string that I used doesn't contain any null string.
And I passed the string with no rank. 

```python
markdown_string =  tf.placeholder(tf.string,name='markdown_string')
```
```python
feed_dict={markdown_string:"Test string contains markdown with table"}
```

And I think current tensorboard code expecting string with no shape.
```python
 if not text_arr.shape: # Here, when there's no shape
  return markdown_and_sanitize(text_arr.astype(np.dtype(str)).tostring())
```

But current conversion using `tostring` add null padding. 

Example code
```python
import tensorflow as tf
import numpy as np 
result = tf.make_ndarray(tf.make_tensor_proto('Test String', dtype=tf.string)).astype(np.dtype(str)).tostring().decode('utf-8')
print("Original Conversion using tostring")
print(repr(result))

result = np.array_str(tf.make_ndarray(tf.make_tensor_proto('Test String', dtype=tf.string)).astype(np.dtype(str)))
print("\nConversion using np.array_str")
print(repr(result))
```

result
```
Original Conversion using tostring
'T\x00\x00\x00e\x00\x00\x00s\x00\x00\x00t\x00\x00\x00 \x00\x00\x00S\x00\x00\x00t\x00\x00\x00r\x00\x00\x00i\x00\x00\x00n\x00\x00\x00g\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'

Conversion using np.array_str
'Test String'
```

So...
* the contents of the string tensor that you're passing to tf.summary.text contains NUL characters;

is false.

And I think these are all false. 

* the event files on disk contain NUL characters within your text;
* the protobufs read into memory by TensorBoard's event accumulator contain NUL characters;
* the tensor protos returned by self._multiplexer.Tensors in the text plugin contain NUL characters;
* the result of tf.make_ndarray(event.tensor_proto) contains NUL characters;

But before this one.
* any argument to markdown_and_sanitize contains NUL characters.

We should add 
* converted string using text_arr.astype(np.dtype(str)).tostring() contains NUL characters. 

The code that tensorboard add null padding is `text_arr.astype(np.dtype(str)).tostring()` 
I checked the result before and after this code.
And the code that I changed was that particular code adding NULL padding.

RyanK1m commented:
I added a test case which will work with my fix but not with the current version.

wchargin commented:
I can't reproduce your original problem:
```py
Python 2.7.6 (default, Oct 26 2016, 20:30:19) 
[GCC 4.8.4] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> import numpy as np 
>>> result = tf.make_ndarray(tf.make_tensor_proto('Test String', dtype=tf.string)).astype(np.dtype(str)).tostring().decode('utf-8')
>>> print("Original Conversion using tostring")
Original Conversion using tostring
>>> print(repr(result))
u'Test String'
>>> 
```

wchargin commented:
Furthermore, when I run your modified test on `origin/master` _without_ your change to the plugin, it works fine. [Here's my log.](https://gist.github.com/wchargin/ce98a32aba9a4b7a0988c03a488de7f1) I'm using the 2017-07-26 TensorFlow nightly on Ubuntu 14.04.5 LTS.

RyanK1m commented:
It seems the problem shows up on Python 3 but not Python 2
I just checked both Python 3.6.1 and Python 2.7.6 and I can confirm this bug on Python 3.6.1

nfelt commented:
Duplicate of #685
